File: src/pgc/SVDistribution.py
#!/usr/env/bin python
import collections
import argparse

#usage: python ~/github/TrioWGS/src/TrioWGS/SVDistribution.py --input /home/wuzhikun/Project/NanoTrio/SVStats/Sniffles/minimap2/M628-2_position.xls --summary temp

__author__ = "Zhikun Wu"
__date__ = "2019.04.15"
__email__ = "598466208@qq.com"

def get_SV_infor(sv_bed):
    """
    sv_bed:
    1   76683   76789   DEL
    1   83940   84004   DEL
    1   136275  136340  INS
    """
    ChrTypeNum = collections.defaultdict(lambda: collections.Counter())

    in_h = open(sv_bed, "r")
    for line in in_h:
        lines = line.strip().split("\t")
        Chr, Satrt, End, Type = lines[:4]
        ChrTypeNum[Chr][Type] += 1

    in_h.close()
    return ChrTypeNum

def SV_distribution(sv_bed, summary_file, chroms, types):
    ChrTypeNum = get_SV_infor(sv_bed)

    su_h = open(summary_file, "w")
    Chroms = chroms.split(",")
    Chroms = [c.strip() for c in Chroms]
    Types = types.split(",")
    Types = [t.strip() for t in Types]
    su_h.write("Chr\t%s\tTotal\n" % "\t".join(Types))
    for Chr in Chroms:
        if Chr in ChrTypeNum:
            typeNum = []
            for t in Types:
                if t in ChrTypeNum[Chr]:
                    number = ChrTypeNum[Chr][t]
                else:
                    number = 0
                typeNum.append(number)
            totalNum = sum(typeNum)
            typeNum = [str(t) for t in typeNum]
            su_h.write("%s\t%s\t%d\n" % (Chr, "\t".join(typeNum), totalNum))


def main():
    parser = argparse.ArgumentParser(description="Get the summary of structural variants distribution in chromosome.")
    parser.add_argument("-i", "--input", help="The input file with SV position and type.")
    parser.add_argument("-s", "--summary", help="The summary of SV distribution.")
    parser.add_argument("-c", "--chromosome", default="1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y", help="The target chromosomes which are separated with comma.")
    parser.add_argument("-t", "--type", default="DEL,INS,DUP,INV,TRA,INVDUP", help="The target SV types.")
    args = parser.parse_args()
    SV_distribution(args.input, args.summary, chroms=args.chromosome, types=args.type)

if __name__ == "__main__":
    main()

File: script/SVDistribution.R
#!/usr/bin/env Rscript
# library(CMplot)
### modified CMplot.r
source("/home/wuzhikun/github/NanoHub/script/CMplot.r") 
library(argparser)


### package
### https://cloud.r-project.org/src/contrib/CMplot_3.3.3.tar.gz

#usage: Rscript ~/github/TrioWGS/script/SVDistribution.R --input /home/wuzhikun/Project/NanoTrio/SVStats/Sniffles/minimap2/M628-2_position.xls --pdf /home/wuzhikun/Project/NanoTrio/SVStats/Sniffles/minimap2/M628-2_position.pdf


arg <- arg_parser('Density plot for structural variant types.')
arg <- add_argument(arg, '--input', help='The file with type and number of SV.')
arg <- add_argument(arg, '--pdf', help='output file with pdf format.')
arg <- add_argument(arg, '--maxNum', help='The max number for heatmap.')
argv <- parse_args(arg)



SV_distribution <- function(in_file, out_file, maxNum){
    ### in_file
    # 1       76683   76789   DEL
    # 1       83940   84004   DEL
    # 1       136275  136340  INS
    # 1       136905  137040  INS


    SV_pos <- read.table(in_file, sep="\t", header=FALSE)
    colnames(SV_pos) <- c("Chromosome", "Position", "Position2", "Type")
    SV_pos$SNP <- paste(SV_pos$Chromosome, SV_pos$Position, SV_pos$Position2, sep="_")

    SV_position <- data.frame(SNP=SV_pos$SNP, Chromosome=SV_pos$Chromosome, Position=SV_pos$Position, Position2=SV_pos$Position2)

    # chrom <- "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y"
    # chromAll <- gsub(",", "|", chrom)
    # chroms <- strsplit(chrom, split=",")
    chroms <- c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22, "X")

    New_position <- SV_position[SV_position$Chromosome %in% chroms,]

    maxNum <- as.numeric(maxNum)
    CMplot(New_position, outName=out_file, maxNum=maxNum, plot.type="d",bin.size=1e6,col=c("darkgreen", "gold2", "red"),file="pdf",memo="",dpi=300, file.output=TRUE, verbose=TRUE, chr.labels=NULL)

}


SV_distribution(argv$input, argv$pdf, argv$maxNum)
